### PR TITLE
Add cleaning.py from Joshuab

### PR DESCRIPTION
**Pull request by jpmcga**
First attempt at re-writing cleaning in python

### CODE REVIEW

File: wf/clean.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    return combined_table
+
+def clean_fragments(fragments_path, r_table):
+    # Reduce high tixels by randomly downsampling fragments.tsv according to
+    # reduction table
+    skip_headers = [i+1 for i in range(50)]
+    print("Loading fragments.tsv")
+    # fragments = gzip.open(fragments_path)
+    fragments = pd.read_csv(fragments_path, sep='\t', header=skip_headers, skiprows=skip_headers)
+    fragments.columns = ['V1', 'V2', 'V3', 'barcode', 'V4']
```
Comment (jpmcga) [inline wf/clean.py:144]: Looks like using header=skip_headers actually moves the first 50 rows into the header; try, 
```suggestion
fragments = pd.read_csv(fragments_path, sep='\t', header=None, skiprows=list(range(51)))
fragments.columns = ['V1', 'V2', 'V3', 'barcode', 'V4']
```

File: wf/clean.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    barcode_groups = frag_copy.groupby('barcode')
+    list_concat = []
+    for i in outlier_barcodes:
+      unique_barcode = barcode_groups.get_group(i)
+      if unique_barcode.shape[0] > int(r_table[i]):
+        unique_barcode = unique_barcode.sample(n= int(r_table[i]))
+      list_concat.append(unique_barcode)
+    
+    downsampled_barcode = pd.concat(list_concat)
+    fragments_cleaned = pd.concat([downsampled_barcode, fragments])
```
Comment (jpmcga) [inline wf/clean.py:160]: This is a clever way to do this.  We don't need the copy though.  Let's try,
```suggestion
    outlier_barcodes = list(r_table.keys())
    normal_frags = fragments[fragments['barcode'].isin(outlier_barcodes) == False]
    outlier_frags = frag_copy[frag_copy['barcode'].isin(outlier_barcodes) == True]
    
    # To each df in the list, randomly downsample if in reduction list
    print("Downsampling....")
    barcode_groups = outlier_frags.groupby('barcode')
    list_concat = []
    for i in outlier_barcodes:
      outlier = barcode_groups.get_group(i)
      if outlier.shape[0] > int(r_table[i]):
        outlier = outlier.sample(n= int(r_table[i]))
      list_concat.append(outlier)
    
    downsampled_frags = pd.concat(list_concat)
    fragments_cleaned = pd.concat([downsampled_frags, normal_frags])
```

File: wf/clean.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    return fragments_cleaned
+
+# do stuff ---------------------------------------------------------------------
+
```
Comment (jpmcga) [inline wf/clean.py:165]: @JoshuaBAtlas let's stick this in a 
```
if __name__ == '__main__':
```

**Review by jpmcga — COMMENTED** [review]
@JoshuaBAtlas take a look at my feedback.  It's mostly regarding following PEP 8.  Let's try to use PEP 8 as much as possible in our collaborative code.  Can you please implement and commit the feedback?  If you switch me to 'Assignee' when you're done, I'll go through it again.  Once clean.py is ready, I'll show you how to swap it out in __init__.py.  I think we can just keep adding commits to this MR--but pls correct me it's better practice to merge this branch and start a new one for implementing.

Additionally, can you provide evidence that the python and R scripts provide the same outputs (ie. cleaned_fragments.tsv's from produced by both scripts).  Functionally, the code looks good to me so I'd expect they'd be the same.

File: .DS_Store
Comment (jpmcga) [inline .DS_Store:1]: @JoshuaBAtlas delete pls

**Review by JoshuaBAtlas — APPROVED** [review]

**Pull request merged**